File: frontend/src/services/radiologyService.ts
import type { ImagingStudy, RadiologyReport } from '../types/radiology';
import api from './api';

export const RadiologyService = {
  // Create a new study
  createStudy: async (study: Partial<ImagingStudy>): Promise<ImagingStudy> => {
    const response = await api.post<ImagingStudy>('/radiology/studies', study);
    return response.data;
  },

  // List studies with filtering
  listStudies: async (
    page = 1,
    limit = 20,
    patientId?: number,
    status?: string
  ) => {
    const params: any = { page, limit };
    if (patientId) params.patient_id = patientId;
    if (status) params.status = status;

    const response = await api.get<{ data: ImagingStudy[]; total: number }>(
      '/radiology/studies',
      { params }
    );
    return response.data;
  },

  // Get study by ID
  getStudy: async (id: number): Promise<ImagingStudy> => {
    const response = await api.get<ImagingStudy>(`/radiology/studies/${id}`);
    return response.data;
  },

  // Update study status
  updateStatus: async (
    id: number,
    status: string,
    techId?: number
  ): Promise<void> => {
    await api.put(`/radiology/studies/${id}/status`, {
      status,
      tech_id: techId,
    });
  },

  // Get worklist
  getWorklist: async (): Promise<ImagingStudy[]> => {
    const response = await api.get<ImagingStudy[]>('/radiology/worklist');
    return response.data;
  },

  // Create report
  createReport: async (
    report: Partial<RadiologyReport>
  ): Promise<RadiologyReport> => {
    const response = await api.post<RadiologyReport>(
      '/radiology/reports',
      report
    );
    return response.data;
  },

  // Update report
  updateReport: async (
    id: number,
    report: Partial<RadiologyReport>
  ): Promise<RadiologyReport> => {
    const response = await api.put<RadiologyReport>(
      `/radiology/reports/${id}`,
      report
    );
    return response.data;
  },
};

File: backend/internal/service/clinical/radiology_service.go
package service

import (
	"errors"
	"fmt"
	"time"

	"github.com/zarishsphere/zarish-his/internal/models"
	"github.com/zarishsphere/zarish-his/internal/repository"
)

type RadiologyService struct {
	repo *repository.RadiologyRepository
}

func NewRadiologyService(repo *repository.RadiologyRepository) *RadiologyService {
	return &RadiologyService{repo: repo}
}

func (s *RadiologyService) CreateStudy(study *models.ImagingStudy) error {
	// Generate UIDs if not present (simplified for MVP)
	if study.StudyUID == "" {
		study.StudyUID = fmt.Sprintf("1.2.840.10008.%d", time.Now().UnixNano())
	}
	if study.AccessionNumber == "" {
		study.AccessionNumber = fmt.Sprintf("ACC-%d", time.Now().Unix())
	}

	study.Status = "scheduled"
	study.StartedAt = time.Now()

	return s.repo.CreateStudy(study)
}

func (s *RadiologyService) GetStudy(id uint) (*models.ImagingStudy, error) {
	return s.repo.GetStudy(id)
}

func (s *RadiologyService) ListStudies(patientID uint, status string, page, limit int) ([]models.ImagingStudy, int64, error) {
	offset := (page - 1) * limit
	return s.repo.ListStudies(patientID, status, limit, offset)
}

func (s *RadiologyService) StartExam(id uint, techID uint) error {
	study, err := s.repo.GetStudy(id)
	if err != nil {
		return err
	}

	if study.Status != "scheduled" {
		return errors.New("study can only be started from scheduled status")
	}

	study.Status = "in-progress"
	study.PerformingTechID = &techID

	// Use repository update or save
	// For simplicity, we'll use the status update method but ideally we'd update multiple fields
	return s.repo.UpdateStudyStatus(id, "in-progress")
}

func (s *RadiologyService) CompleteExam(id uint) error {
	// In a real system, we'd verify that images have been received
	// Also update CompletedAt in a real implementation
	return s.repo.UpdateStudyStatus(id, "completed")
}

func (s *RadiologyService) CreateReport(report *models.RadiologyReport) error {
	// Check if study exists
	study, err := s.repo.GetStudy(report.StudyID)
	if err != nil {
		return err
	}

	if study.Status != "completed" {
		return errors.New("cannot report on an incomplete study")
	}

	report.ReportedAt = time.Now()
	if report.Status == "" {
		report.Status = "draft"
	}

	return s.repo.CreateReport(report)
}

func (s *RadiologyService) FinalizeReport(id uint) error {
	// This would fetch the report, check permissions, and update status
	// For MVP, we'll assume the handler passes a report object with status="final" to UpdateReport
	return nil
}

func (s *RadiologyService) UpdateReport(report *models.RadiologyReport) error {
	if report.Status == "final" {
		now := time.Now()
		report.FinalizedAt = &now
	}
	return s.repo.UpdateReport(report)
}

func (s *RadiologyService) GetWorklist() ([]models.ImagingStudy, error) {
	// Radiologist worklist: completed studies needing reports
	// Technician worklist: scheduled studies
	// For now, return all active
	return s.repo.GetWorklist([]string{"scheduled", "in-progress", "completed"})
}
